In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
clicks = pd.read_csv('./datos/clicks.csv')

In [3]:
clicks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26351 entries, 0 to 26350
Data columns (total 20 columns):
advertiser_id      26351 non-null int64
action_id          0 non-null float64
source_id          26351 non-null int64
created            26351 non-null object
country_code       26351 non-null int64
latitude           26351 non-null float64
longitude          26351 non-null float64
wifi_connection    26351 non-null bool
carrier_id         26340 non-null float64
trans_id           26351 non-null object
os_minor           26339 non-null float64
agent_device       3243 non-null float64
os_major           26339 non-null float64
specs_brand        26351 non-null int64
brand              6235 non-null float64
timeToClick        22977 non-null float64
touchX             23011 non-null float64
touchY             23011 non-null float64
ref_type           26351 non-null int64
ref_hash           26351 non-null int64
dtypes: bool(1), float64(11), int64(6), object(2)
memory usage: 3.8+ MB


In [6]:
clicks.sample(5)

advertiser_id  action_id  source_id                   created  \
4584               3        NaN          0  2019-03-09T01:30:06.126Z   
11812              3        NaN          0  2019-03-11T03:34:39.691Z   
2319               3        NaN          5  2019-03-08T01:42:57.142Z   
3634               3        NaN          0  2019-03-10T12:49:20.145Z   
7281               3        NaN          6  2019-03-09T04:42:22.879Z   

              country_code  latitude  longitude  wifi_connection  carrier_id  \
4584   6333597102633388268  1.217715   1.064799            False         4.0   
11812  6333597102633388268  1.205689   1.070234            False        17.0   
2319   6333597102633388268  1.205689   1.070234            False        13.0   
3634   6333597102633388268  1.205689   1.070234            False         4.0   
7281   6333597102633388268  1.206684   1.071447            False        13.0   

                              trans_id      os_minor  agent_device  \
4584   ZdhpthwAcls3zYTmveDquUrdL4iDAik  6.768137e+18           NaN   
11812  i4wthpO8P6ez8v0X-jQVGBMbyOWim1g  3.511538e+18           NaN   
2319   BPT3-9xyFLsA_CB3uBCRPlI6gKm9ymA  3.378640e+18           NaN   
3634   FwooqS0vyF8Js3s5MeUjLg97-CwD08c  5.975656e+18           NaN   
7281   JRU4V-hoKgBXJzpYHiQYM2Et-tNPGs4  6.768137e+18           NaN   

           os_major          specs_brand  brand  timeToClick  touchX  touchY  \
4584   3.072849e+18   392184377613098015    NaN      499.755   0.807   0.507   
11812  1.774085e+18   392184377613098015    NaN        5.201   0.773   0.692   
2319   4.566454e+18  2733035977666442898    NaN          NaN     NaN     NaN   
3634   3.072849e+18   392184377613098015    NaN        1.226   0.749   0.500   
7281   3.072849e+18  2733035977666442898    NaN        1.813   0.887   0.133   

                  ref_type             ref_hash  
4584   1891515180541284343  5096882600526921860  
11812  1891515180541284343   639817810790733548  
2319   1891515180541284343  4018049520811404462  
3634   1891515180541284343  9109571991877350683  
7281   1891515180541284343  7709815538947523480

In [4]:
clicks.shape

(26351, 20)

## A continuación se analizan los valores que contienen las respectivas columnas 

Primero se analizan la cantidad de nulos que se tienen en el archivo.

Como podemos ver en la siguiente salida, la columa 'action_id' tiene una cantidad de 26351 nulos. A su vez, el archivo clicks.csv posee una cantidad de 26351 filas, por lo que no hay niguna fila de 'action_id' que contenga algún valor no nulo.

In [7]:
clicks.isnull().sum()

advertiser_id          0
action_id          26351
source_id              0
created                0
country_code           0
latitude               0
longitude              0
wifi_connection        0
carrier_id            11
trans_id               0
os_minor              12
agent_device       23108
os_major              12
specs_brand            0
brand              20116
timeToClick         3374
touchX              3340
touchY              3340
ref_type               0
ref_hash               0
dtype: int64

Si analizamos la cantidad de valores diferentes que se tienen en la columna de 'country_code', a través de la función value_count(), podemos determinar que todos los clicks registrados son del mismo país, por lo que no es conveniente utilizar dicho columna para un posterior análisis. Por ende, eliminamos la columna en cuestión para reducir el tamaño del archivo. 

In [10]:
clicks['country_code'].value_counts()

6333597102633388268    26351
Name: country_code, dtype: int64

In [11]:
clicks = clicks.drop(['country_code'], axis='columns')

In [12]:
clicks.sample(5)

advertiser_id  action_id  source_id                   created  \
14483              3        NaN          0  2019-03-11T15:49:10.161Z   
11988              3        NaN          0  2019-03-11T02:10:00.988Z   
18108              3        NaN          6  2019-03-13T02:37:35.466Z   
11660              3        NaN          0  2019-03-11T03:50:48.777Z   
318                3        NaN          0  2019-03-07T23:12:41.633Z   

       latitude  longitude  wifi_connection  carrier_id  \
14483  1.205689   1.070234            False         4.0   
11988  1.205689   1.070234            False         4.0   
18108  1.205689   1.070234            False        13.0   
11660  1.205689   1.070234            False        12.0   
318    1.205689   1.070234            False         6.0   

                              trans_id      os_minor  agent_device  \
14483  qEeNWOLdQAwWbOZpccbRCcxi929r9YI  6.768137e+18           NaN   
11988  2fjk4qeAs7x_lIWBBRcpDYjcQij60ew  3.378640e+18           NaN   
18108  OjDbyJs9Vxizv0usL6LOWHeL28aVjd8  6.391567e+18           NaN   
11660  JrKoScx55MWj7OEoxCwmmbvMc0eixyo  6.768137e+18           NaN   
318    CmKkjjBMGYovto_KBHQowCr25YbWkJo  6.391567e+18           NaN   

           os_major          specs_brand  brand  timeToClick  touchX  touchY  \
14483  3.072849e+18   392184377613098015    2.0       27.646   0.498   0.659   
11988  4.566454e+18   392184377613098015    NaN      148.513   0.810   5.428   
18108  7.452406e+18  2733035977666442898    NaN          NaN     NaN     NaN   
11660  3.072849e+18   392184377613098015    NaN      529.043   0.178   7.158   
318    7.452406e+18   392184377613098015    NaN        2.071   0.557   0.073   

                  ref_type             ref_hash  
14483  1891515180541284343  1179074484659087586  
11988  1891515180541284343  3663517142061027018  
18108  1891515180541284343  5397135629481185689  
11660  1891515180541284343  6529998342038747288  
318    1891515180541284343  3397040402906991429

Para las columnas 'latitude' y 'longitude' podemos cambiarle el tipo para achicar el tamaño a manejar. Pasamos de float64 a float32. 

In [13]:
clicks['latitude'].head()

0    1.205689
1    1.218924
2    1.205689
3    1.205689
4    1.205689
Name: latitude, dtype: float64

In [14]:
clicks['longitude'].head()

0    1.070234
1    1.071209
2    1.070234
3    1.070234
4    1.070234
Name: longitude, dtype: float64

In [15]:
clicks['latitude'] = clicks['latitude'].astype('float32')
clicks['longitude'] = clicks['longitude'].astype('float32')

Para la columna 'created' le podemos cambiar el tipo a 'datetime'. A su vez, separamos dicha fecha a las columnas 'year', 'month', 'day' y 'hour'.

In [17]:
clicks['created'] = pd.to_datetime(clicks['created'])

In [18]:
clicks['created'].head()

0   2019-03-06 22:42:12.755
1   2019-03-08 10:24:30.641
2   2019-03-08 15:24:16.069
3   2019-03-06 03:08:51.543
4   2019-03-06 03:32:55.570
Name: created, dtype: datetime64[ns]

In [36]:
clicks['year'] = clicks['created'].dt.year
clicks['month'] = clicks['created'].dt.month
clicks['day'] = clicks['created'].dt.day
clicks['hour'] = clicks['created'].dt.hour

In [37]:
clicks.sample(5)

advertiser_id  action_id  source_id                 created  latitude  \
937                3        NaN          0 2019-03-08 01:51:13.644  1.205689   
25020              3        NaN          0 2019-03-12 01:59:48.989  1.205689   
17104              3        NaN          0 2019-03-10 01:48:46.456  1.205689   
17650              3        NaN          0 2019-03-10 06:01:25.336  1.205689   
19033              3        NaN          1 2019-03-13 17:26:04.656  1.205689   

       longitude  wifi_connection  carrier_id  \
937     1.070234            False         6.0   
25020   1.070234            False         3.0   
17104   1.070234            False         4.0   
17650   1.070234            False         6.0   
19033   1.070234            False         9.0   

                              trans_id      os_minor  ...   brand  \
937    jnXu5cfSudRwnTLxUSw4QzowpktUTpQ  6.391567e+18  ...     NaN   
25020  lCGGPVQ6eBMTe0gtD4V8TRem9H4f_wA  3.511538e+18  ...     NaN   
17104  JaUN-grY9kPglpONzU5yvq1sr1dmff8  5.975656e+18  ...     NaN   
17650  xHlD7PrGtTmyJ8ISjmIGZSR5O7QdRg0  6.391567e+18  ...     NaN   
19033  DpfUx-bfSCU-udAM_3mW58oDk8J1TeA  3.378640e+18  ...     0.0   

       timeToClick  touchX  touchY             ref_type             ref_hash  \
937        836.662   0.752   0.769  1891515180541284343  4647505260741085617   
25020        5.733   0.614   0.783  1891515180541284343  3813322271069091211   
17104       27.021   0.910   0.481  1891515180541284343  4382487184740205547   
17650    14689.199   0.830   0.686  1891515180541284343  8834343268567161273   
19033        0.267   0.488   0.767  1891515180541284343  3630850984945343218   

       year  month  day  hour  
937    2019      3    8     1  
25020  2019      3   12     1  
17104  2019      3   10     1  
17650  2019      3   10     6  
19033  2019      3   13    17  

[5 rows x 23 columns]

Si analizamos detalladamente las columnas creadas anteriormente, podemos observar que tanto para 'year' como para 'month' tenemos solo un valor posible para dichos campos, por lo que podemos eliminar dichas columnas. 

En este caso, todos los clicks registrados corresponden al mes de Marzo del 2019.

In [38]:
clicks['year'].value_counts()

2019    26351
Name: year, dtype: int64

In [39]:
clicks['month'].value_counts()

3    26351
Name: month, dtype: int64

In [40]:
clicks = clicks.drop(['year'], axis='columns')
clicks = clicks.drop(['month'], axis='columns')

Para el caso de 'day' y 'hour', verificamos que tenemos diferentes valores pero están seteados como del tipo int64. Como vamos a tener enteros pero valores chicos, podemos tranformar el tipo a int8.

In [49]:
clicks['day'].value_counts()

12    5204
11    4557
10    3966
13    3468
9     3283
7     2989
8     2839
5       31
6       14
Name: day, dtype: int64

In [42]:
clicks['hour'].value_counts()

1     4209
23    2726
2     2327
11    1455
22    1386
12    1199
10    1143
4     1079
16     966
15     959
5      945
13     926
14     834
9      754
3      754
6      678
18     655
17     641
8      537
7      518
21     451
0      440
20     389
19     380
Name: hour, dtype: int64

In [52]:
clicks['day'] = clicks['day'].astype('int8')
clicks['hour'] = clicks['hour'].astype('int8')

In [53]:
clicks['carrier_id'].head()

0     1.0
1     4.0
2     6.0
3    45.0
4    45.0
Name: carrier_id, dtype: float64

In [54]:
clicks['carrier_id'].apply(lambda x: x % 1).sum()

0.0

In [55]:
clicks['carrier_id'].max()

116.0

In [56]:
clicks['carrier_id'].min()

0.0

In [57]:
clicks['carrier_id'] = clicks['carrier_id'].astype('int8')

ValueError: Cannot convert non-finite values (NA or inf) to integer